storage: TestStoreRangeUpReplicate failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/99ba774d014955965f819d36fac7e14b2d405aa9

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeUpReplicate PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1490198&tab=buildLog

```
I190916 06:21:16.429299 165300 storage/replica_raft.go:291  [replicate,s1,r18/1:/Table/2{2-3},txn=5fee10b4] proposing ADD_REPLICA[(n3,s3):3]: after=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190916 06:21:16.433140 172856 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190916 06:21:16.433196 172856 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190916 06:21:16.433505 172854 util/stop/stopper.go:542  quiescing; tasks left:
47     [async] storage.IntentResolver: cleanup txn records
1      storage.replicate: purgatory processing replica
1      storage.IntentResolver: processing intents
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190916 06:21:16.433611 172855 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190916 06:21:16.433706 172855 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190916 06:21:16.433753 172854 util/stop/stopper.go:542  quiescing; tasks left:
47     [async] storage.IntentResolver: cleanup txn records
1      storage.replicate: purgatory processing replica
1      storage.IntentResolver: processing intents
1      [async] intent_resolver_gc_batcher
E190916 06:21:16.433762 165681 storage/queue.go:1027  [replicate,s1,r6/1:/Table/{SystemCon…-11}] change replicas of r6 failed: fetching current range descriptor value: node unavailable; try another peer
I190916 06:21:16.433784 165681 storage/queue.go:1127  [replicate] purgatory is now empty
I190916 06:21:16.433826 172854 util/stop/stopper.go:542  quiescing; tasks left:
47     [async] storage.IntentResolver: cleanup txn records
1      storage.IntentResolver: processing intents
1      [async] intent_resolver_gc_batcher
W190916 06:21:16.433830 170884 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/23/RangeDescriptor: node unavailable; try another peer
I190916 06:21:16.433868 172854 util/stop/stopper.go:542  quiescing; tasks left:
46     [async] storage.IntentResolver: cleanup txn records
1      storage.IntentResolver: processing intents
1      [async] intent_resolver_gc_batcher
I190916 06:21:16.433926 172854 util/stop/stopper.go:542  quiescing; tasks left:
46     [async] storage.IntentResolver: cleanup txn records
1      storage.IntentResolver: processing intents
I190916 06:21:16.433985 172854 util/stop/stopper.go:542  quiescing; tasks left:
45     [async] storage.IntentResolver: cleanup txn records
1      storage.IntentResolver: processing intents
I190916 06:21:16.434624 172854 util/stop/stopper.go:542  quiescing; tasks left:
4      [async] storage.IntentResolver: cleanup txn records
I190916 06:21:16.434669 172854 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
I190916 06:21:16.434713 172854 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
W190916 06:21:16.435632 166556 storage/raft_transport.go:294  unable to accept Raft message from (n3,s3):3LEARNER: no handler registered for (n1,s1):1
I190916 06:21:16.436023 172853 util/stop/stopper.go:542  quiescing; tasks left:
1      rpc heartbeat
W190916 06:21:16.436997 166429 storage/raft_transport.go:620  while processing outgoing Raft queue to node 1: EOF:
W190916 06:21:16.437049 166550 storage/raft_transport.go:620  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W190916 06:21:16.437077 166969 storage/raft_transport.go:620  while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = grpc: the client connection is closing:
--- FAIL: TestStoreRangeUpReplicate (1.93s)
    require.go:157: 
        	Error Trace:	client_raft_test.go:1576
        	Error:      	Not equal: 
        	            	expected: 48
        	            	actual  : 46
        	Test:       	TestStoreRangeUpReplicate

```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -1535,23 +1535,36 @@ func TestStoreRangeUpReplicate(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// The range should become available on every node.
-	var r *storage.Replica // from the last store
+	// Wait until all ranges are upreplicated to all nodes.
+	var replicaCount int64
 	testutils.SucceedsSoon(t, func() error {
-		for _, s := range mtc.stores {
-			r = s.LookupReplica(roachpb.RKey("a"))
-			if r == nil {
-				return errors.Errorf("expected replica for 'a'")
+		var replicaCounts [3]int64
+		for i, s := range mtc.stores {
+			var err error
+			mtc.stores[i].VisitReplicas(func(r *storage.Replica) bool {
+				replicaCounts[i]++
+				// Synchronize with the replica's raft processing goroutine.
+				r.RaftLock()
+				defer r.RaftUnlock()
+				if len(r.Desc().InternalReplicas) != 3 {
+					// This fails even after the snapshot has arrived and only
+					// goes through once the replica has applied the conf change.
+					err = errors.Errorf("not fully initialized")
+					return false
+				}
+				return true
+			})
+			if err != nil {
+				return err
+			}
+			if replicaCounts[i] != replicaCounts[0] {
+				return errors.Errorf("not fully upreplicated")
 			}
 			if n := s.ReservationCount(); n != 0 {
 				return errors.Errorf("expected 0 reservations, but found %d", n)
 			}
-			if len(r.Desc().InternalReplicas) != 3 {
-				// This fails even after the snapshot has arrived and only goes through
-				// once the replica has applied the conf change.
-				return errors.Errorf("not fully initialized")
-			}
 		}
+		replicaCount = replicaCounts[0]
 		return nil
 	})
 
@@ -1566,11 +1579,6 @@ func TestStoreRangeUpReplicate(t *testing.T) {
 	if generated == 0 {
 		t.Fatalf("expected at least 1 snapshot, but found 0")
 	}
-	var replicaCount int64
-	mtc.stores[0].VisitReplicas(func(_ *storage.Replica) bool {
-		replicaCount++
-		return true
-	})
 	// We upreplicate each range (once each for n2 and n3), so there should be
 	// exactly 2 * replica learner snaps, one per upreplication.
 	require.Equal(t, 2*replicaCount, learnerApplied)

